app/aboot/fs_boot.c

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include <debug.h>
+#include <target.h>
+#include <string.h>
+
+#include <lib/bio.h>
+#include <lib/fs.h>
+
+#include "fs_boot.h"
+
+static const char *bootable_parts[] = {
+	"system",
+	"cache",
+	"userdata",
+};
+
+static bool fsboot_bootable_part(char *label)
+{
+	int i;
+	for (i = 0; i < ARRAY_SIZE(bootable_parts); ++i)
+		if (strcmp(label, bootable_parts[i]) == 0)
+			return true;
+
+	return false;
+}
+
+/* Only iterate through files if target is NULL */
+static int fsboot_fs_load_img(char *dev_name, void* target, size_t sz)
+{
+	struct dirhandle *dirh;
+	struct dirent dirent;
+
+	char image_path[128] = "/mnt/";
+	int ret = 0, error = -2;
+
+	if (fs_mount("/mnt", "ext2", dev_name) >= 0) {
+		error = 0;
+		ret = fs_open_dir("/mnt", &dirh);
+		if (ret >= 0) {
+			while (fs_read_dir(dirh, &dirent) >= 0) {
+				if (!target) {
+					dprintf(SPEW, "| /%s/%s\n", dev_name, dirent.name);
+				} else if (strncmp(dirent.name, "boot.img", 7) == 0) {
+					strcpy(&image_path[strlen("/mnt/")], dirent.name);
+					dprintf(INFO, "Found boot image: %s : %s\n", dev_name, image_path);
+					fs_close_dir(dirh);
+					ret = fs_load_file(image_path, target, sz);
+					fs_unmount("/mnt");
+					return ret;
+				}
+			}
+			fs_close_dir(dirh);
+		} else {
+			dprintf(SPEW, "fs_open_dir ret = %d\n", ret);
+			if (target)
+				return ret;
```

Comment:
fs_unmount() missing here. Perhaps do a goto label?

```c
if (fs_mount("/mnt", "ext2", dev_name) < 0)
	return -1;

ret = fs_open_dir("/mnt", &dirh);
// ...

out:
	fs_unmount("/mnt");
	return ret;
```

This would also reduce the indentation level a bit, which I would appreciate quite a lot.